### PR TITLE
[Bugfix] Fix gat residual bug

### PR DESCRIPTION
**Pull request by jermainewang**
## Description

Fix the residual shape problem in #354.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

### CODE REVIEW

**Comment by mufeili** [comment]
In fact, comparing to [the official TensorFlow implementation](https://github.com/PetarV-/GAT/blob/master/utils/layers.py), I think there are two differences:
1. [L56, dgl](https://github.com/dmlc/dgl/blob/master/examples/pytorch/gat/train.py#L56) should be `self.residual_fc = nn.Linear(in_dim, out_dim, bias=False)` according to [L30, official](https://github.com/PetarV-/GAT/blob/master/utils/layers.py#L30)
2. An additional aggregation across heads (concatenation and sum) as in [L15, L23, L29, official](https://github.com/PetarV-/GAT/blob/master/models/gat.py#L15) is missing, which should be placed around [L133, L135, PyTorch](https://github.com/dmlc/dgl/blob/master/examples/pytorch/gat/train.py#L133).

You may also want to fix [L57](https://github.com/dmlc/dgl/blob/master/examples/pytorch/gat/train.py#L57) where `self.fc.weight.data` should probably be `self.residual_fc.weight.data`.

**Comment by mufeili** [comment]
@jermainewang In fact, comparing to [the official TensorFlow implementation](https://github.com/PetarV-/GAT/blob/master/utils/layers.py), I think there are two differences:
1. [L56, dgl](https://github.com/dmlc/dgl/blob/master/examples/pytorch/gat/train.py#L56) should be `self.residual_fc = nn.Linear(in_dim, out_dim, bias=False)` according to [L30, official](https://github.com/PetarV-/GAT/blob/master/utils/layers.py#L30)
2. An additional aggregation across heads (concatenation and sum) as in [L15, L23, L29, official](https://github.com/PetarV-/GAT/blob/master/models/gat.py#L15) is missing, which should be placed around [L133, L135, PyTorch](https://github.com/dmlc/dgl/blob/master/examples/pytorch/gat/train.py#L133).

You may also want to fix [L57](https://github.com/dmlc/dgl/blob/master/examples/pytorch/gat/train.py#L57) where `self.fc.weight.data` should probably be `self.residual_fc.weight.data`.

**Comment by jermainewang** [comment]
Done.

**Comment by jermainewang** [comment]
I guess dividing by the number of heads does not matter? It just applies a scaling and does not affect the softmax?

**Comment by mufeili** [comment]
> I guess dividing by the number of heads does not matter? It just applies a scaling and does not affect the softmax?

Well, it's still different if you plug the logits into the loss function.

File: examples/pytorch/gat/train.py [hunk 1]
```diff
@@ -53,37 +53,38 @@ def __init__(self,
         self.residual = residual
         if residual:
             if in_dim != out_dim:
-                self.residual_fc = nn.Linear(in_dim, num_heads * out_dim, bias=False)
-                nn.init.xavier_normal_(self.fc.weight.data, gain=1.414)
+                self.res_fc = nn.Linear(in_dim, num_heads * out_dim, bias=False)
```
Comment (mufeili) [inline examples/pytorch/gat/train.py]: While this may add more model capacity, I thought they did `nn.Linear(in_dim, out_dim, bias=False)` as in [L30](https://github.com/PetarV-/GAT/blob/master/utils/layers.py#L30)?

File: examples/pytorch/gat/train.py [hunk 2]
```diff
@@ -53,37 +53,38 @@ def __init__(self,
...
                           [fn.sum('ft', 'ft'), fn.sum('a', 'z')])
         # 3. apply normalizer
-        ret = self.g.ndata['ft'] / self.g.ndata['z']
+        ret = self.g.ndata['ft'] / self.g.ndata['z']  # NxHxD'
         # 4. residual
         if self.residual:
-            if self.residual_fc:
-                ret = self.residual_fc(h) + ret
+            if self.res_fc is not None:
+                resval = self.res_fc(h).reshape((h.shape[0], self.num_heads, -1))  # NxHxD'
```
Comment (mufeili) [inline examples/pytorch/gat/train.py]: If my previous comment on `self.res_fc` is correct, this should probably be `self.res_fc(h).unsqueeze(1)`.
